aten/src/ATen/native/mkldnn/Conv.cpp

```diff
@@ -102,121 +102,45 @@ Tensor _mkldnn_conv2d(
 }
 
 at::Tensor mkldnn_convolution(
-    const at::Tensor& input, const at::Tensor& weight, const at::Tensor& bias,
-    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups)
-{
+    const at::Tensor& input,
+    const at::Tensor& weight,
+    const at::Tensor& bias,
+    IntArrayRef padding,
+    IntArrayRef stride,
+    IntArrayRef dilation,
+    int64_t groups) {
+  Tensor mkldnn_input, mkldnn_weight, mkldnn_bias;
   if (input.is_mkldnn()) {
-    return _mkldnn_conv2d(input, weight, bias, padding, stride, dilation, groups);
-  }
-
-  auto output = at::empty(
-      conv_output_size(
-          input.sizes(), weight.sizes(), padding, stride, dilation),
-      input.options());
-
-  auto cpu_engine = CpuEngine::Instance().get_engine();
-
-  int32_t g = groups;
-
-  int32_t n = input.size(0);
-  int32_t ic = input.size(1);
-  int32_t ih = input.size(2);
-  int32_t iw = input.size(3);
-
-  int32_t oc = output.size(1);
-  int32_t oh = output.size(2);
-  int32_t ow = output.size(3);
-
-  int32_t kh = weight.size(2);
-  int32_t kw = weight.size(3);
-
-  int32_t sh = stride[0];
-  int32_t sw = stride[1];
-  int32_t ph = padding[0];
-  int32_t pw = padding[1];
-
-  auto data_t = memory::data_type::f32;
-  auto format_any = memory::format::any;
-  auto format_nchw = memory::format::nchw;
-  auto format_weight = (g!= 1) ? memory::format::goihw : memory::format::oihw;
-  auto format_x = memory::format::x;
-
-  memory::dims input_tz = {n, ic, ih, iw};
-  memory::dims weight_tz = (g!= 1) ? memory::dims{g, oc/g, ic/g, kh, kw} : memory::dims{oc, ic, kh, kw};
-  memory::dims bias_tz = {oc};
-  memory::dims output_tz = {n, oc, oh, ow};
-  memory::dims _stride = {sh, sw};
-  memory::dims _padding = {ph, pw};
-
-  auto input_md = memory::desc({input_tz}, data_t, format_any);
-  auto weight_md = memory::desc({weight_tz}, data_t, format_any);
-  auto bias_md = memory::desc({bias_tz}, data_t, format_any);
-  auto output_md = memory::desc({output_tz}, data_t, format_any);
-
-  std::shared_ptr<convolution_forward::desc> conv_forward_desc;
-  if (bias.defined()) {
-    conv_forward_desc.reset(new convolution_forward::desc(prop_kind::forward,
-      convolution_direct, input_md, weight_md, bias_md, output_md,
-      _stride, _padding, _padding, padding_kind::zero));
+    mkldnn_input = input;
+    AT_ASSERTM(weight.is_mkldnn(), "weight is supposed to be a mkldnn tenser");
+    mkldnn_weight = weight;
+    if (bias.defined()) {
+      AT_ASSERTM(bias.is_mkldnn(), "bias is supposed to be a mkldnn tensor");
+      mkldnn_bias = bias;
+    }
   } else {
-    conv_forward_desc.reset(new convolution_forward::desc(prop_kind::forward,
-      convolution_direct, input_md, weight_md, output_md,
-      _stride, _padding, _padding, padding_kind::zero));
+    mkldnn_input = dense_to_mkldnn(input);
```

Comment:
I guess here is why the code is slower than before. `dense_to_mkldnn` introduces one more copy while originally we just used `data_ptr` of CPU tensor directly for reorder.

Old path: input (CPU NCHW) -> reordered input (MKL-DNN blocked) -> conv
New path: input (CPU NCHW) -> input' (MKL-DNN NCHW) -> reordered input (MKL-DNN blocked) -> conv

@bddppq You may consider to initiate an `ideep::tensor` directly from `input.data_ptr` with init function `void init(const descriptor &adesc, void *ahandle)`:
https://github.com/intel/ideep/blob/3a9867597c43b4508b1fbdd1ecfbc79e62af1571/include/ideep/tensor.hpp#L630

Reply:
@jgong5 Yes I had some discussions with @Jianhui-Li and we found this place is indeed the extra reordering. I think we will need ideep to expose something similar to `ideep::convolution_forward::expected_weights_descriptor` but for conv2d input to get the descriptor - conceptually copy the current `expected_weights_descriptor` and change `return comp.dup_weights_descriptor();` to `return comp.dup_src_descriptor();` at here https://bddppq.github.io/codebrowser/pytorch/pytorch/third_party/ideep/include/ideep/computations.hpp.html#2258 should be sufficient.
Another thing I have noticed is inside `expected_weights_descriptor` it should use `fetch_or_create_m` to cache the computation at https://bddppq.github.io/codebrowser/pytorch/pytorch/third_party/ideep/include/ideep/computations.hpp.html#2255

Reply:
@Jianhui-Li @jgong5 This is what I meant https://github.com/bddppq/ideep/commit/190d1ab6eb4ed2581ba02b697f7367097e07f5ea